### PR TITLE
Test cases for nodeset error handling

### PR DESCRIPTION
**Pull request by neo954**
These are test cases against pull request #4037.
Bug #3838 
GitHub task #4068

### CODE REVIEW

**Comment by robin2008** [comment]
I'm not sure it is a right way to use shell script for all the testing, instead of keep our conventions and write steps in the test case itself.

I know the framework has some limitation, and we have to write some helper scripts to do something. But if the requirement is general, I prefer to enhance the framework, for example, here we may just need a 'including' feature in case level, we can have a new terminology, for example,  `sub-case`, to contains some common steps, and they could be included other cases.

I don't think it would take many effort.

Another enhancement might be to define a variable during the test running life-cycle. It could also make our test case looks more graceful.

File: xCAT-test/autotest/testcase/nodeset/nodeset_nonexistent_osimage_grub2 [hunk 1]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+
+BASE_DIR="${0%/*}"
+
+! source "${BASE_DIR}/functions.sh" >/dev/null 2>&1 &&
+        echo "File \"${BASE_DIR}/functions.sh\" not found" >&2 && exit 1
+
+make_bogus_ppc64_osimage
```
Comment (robin2008) [inline xCAT-test/autotest/testcase/nodeset/nodeset_nonexistent_osimage_grub2]: Better to have more output there as we need to check the xcattest log files to see what's happens.

File: xCAT-test/autotest/testcase/nodeset/nodeset_nonexistent_osimage_grub2 [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+# Make sure all other nodes are good
+for node in tz{001..005}
+do
+	[ -f "/tftpboot/petitboot/${node}" ]
+	exit_if_bad "$?" "file not found /tftpboot/petitboot/${node}" 
+done
+
+nodeset tz003 osimage=nonexistent-osimage
+[ "$?" -ne "0" ]
+exit_if_bad "$?" "nodeset command should exit with non-zero"
```
Comment (robin2008) [inline xCAT-test/autotest/testcase/nodeset/nodeset_nonexistent_osimage_grub2]: we need to check if the right node report the error.

File: xCAT-test/autotest/testcase/nodeset/nodeset_nonexistent_osimage_grub2 [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+nodeset tz001+4 offline
+
+nodeset tz001+4 osimage
+[ "$?" -ne "0" ]
+exit_if_bad "$?" "nodeset command should exit with non-zero"
```
Comment (robin2008) [inline xCAT-test/autotest/testcase/nodeset/nodeset_nonexistent_osimage_grub2]: Need to check the right failure node based on the previous settings.

File: xCAT-test/autotest/testcase/nodeset/nodeset_nonexistent_osimage_petitboot [hunk 4]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
```
Comment (robin2008) [inline xCAT-test/autotest/testcase/nodeset/nodeset_nonexistent_osimage_petitboot]: same issue as above.

File: xCAT-test/autotest/testcase/nodeset/nodeset_nonexistent_osimage_xnba [hunk 5]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
```
Comment (robin2008) [inline xCAT-test/autotest/testcase/nodeset/nodeset_nonexistent_osimage_xnba]: same issue as above.

File: xCAT-test/autotest/testcase/nodeset/nodeset_nonexistent_osimage_petitboot [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+make_bogus_ppc64le_osimage
+make_bogus_petitboot_nodes
+
+function custom_cleanup()
+{
+	destory_bogus_nodes
+	destory_bogus_osimages
+}
+
+nodeset tz001+4 osimage=rhels7.99-ppc64le-install-compute
```
Comment (robin2008) [inline xCAT-test/autotest/testcase/nodeset/nodeset_nonexistent_osimage_petitboot]: I'd like the osimage should be passed in, also for the setup/cleanup.  I don't want a new script is required when testing for sles and ubuntu.

**Comment by neo954** [comment]
Please check below for the unit test result.

```
# xcattest -t nodeset_shell_grub2,nodeset_shell_petitboot,nodeset_shell_xnba,nodeset_osimage_grub2,nodeset_osimage_petitboot,nodeset_osimage_xnba,nodeset_nonexistent_osimage_grub2,nodeset_nonexistent_osimage_petitboot,nodeset_nonexistent_osimage_xnba
xCAT automated test started at Thu Oct 26 02:25:34 2017
******************************
loading test cases
******************************
To run:
nodeset_shell_grub2
nodeset_shell_petitboot
nodeset_shell_xnba
nodeset_osimage_grub2
nodeset_osimage_petitboot
nodeset_osimage_xnba
nodeset_nonexistent_osimage_grub2
nodeset_nonexistent_osimage_petitboot
nodeset_nonexistent_osimage_xnba
******************************
Start to run test cases
******************************
------START::nodeset_shell_grub2::Time:Thu Oct 26 02:25:34 2017------

RUN:/opt/xcat/share/xcat/tools/autotest/testcase/nodeset/nodeset_shell_grub2 [Thu Oct 26 02:25:34 2017]
ElapsedTime:12 sec
RETURN rc = 0
OUTPUT:
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
tz003: Error: No archictecture defined in nodetype table for the node.
Error: Failed to generate grub2 configurations for some node(s) on c910f03c01p09. Check xCAT log file for more details.
tz001: shell
tz002: shell
tz004: shell
tz005: shell
tz003: Error: No archictecture defined in nodetype table for the node.
tz003: Error: No archictecture defined in nodetype table for the node.
Error: Failed to generate grub2 configurations for some node(s) on c910f03c01p09. Check xCAT log file for more details.
tz001: offline
tz002: offline
tz003: offline
tz004: offline
tz005: offline
5 object definitions have been removed.
Error: Could not find an object named 'rhels7.99-ppc64le-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
1 object definitions have been removed.
Error: Could not find an object named 'rhels6.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels6.99-x86_64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
CHECK:rc == 0	[Pass]

------END::nodeset_shell_grub2::Passed::Time:Thu Oct 26 02:25:46 2017 ::Duration::12 sec------
------START::nodeset_shell_petitboot::Time:Thu Oct 26 02:25:46 2017------

RUN:/opt/xcat/share/xcat/tools/autotest/testcase/nodeset/nodeset_shell_petitboot [Thu Oct 26 02:25:46 2017]
ElapsedTime:8 sec
RETURN rc = 0
OUTPUT:
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
5 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
tz003: Error: No archictecture defined in nodetype table for the node.
tz004: Error: serialport defined, but no serialspeed for this node in nodehm table
Error: Failed to generate petitboot configurations for some node(s) on c910f03c01p09. Check xCAT log file for more details.
tz001: shell
tz002: shell
tz005: shell
tz003: Error: No archictecture defined in nodetype table for the node.
tz004: Error: serialport defined, but no serialspeed for this node in nodehm table
tz001: offline
tz002: offline
tz003: offline
tz004: offline
tz005: offline
5 object definitions have been removed.
1 object definitions have been removed.
Error: Could not find an object named 'rhels7.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels6.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels6.99-x86_64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
CHECK:rc == 0	[Pass]

------END::nodeset_shell_petitboot::Passed::Time:Thu Oct 26 02:25:54 2017 ::Duration::8 sec------
------START::nodeset_shell_xnba::Time:Thu Oct 26 02:25:54 2017------

RUN:/opt/xcat/share/xcat/tools/autotest/testcase/nodeset/nodeset_shell_xnba [Thu Oct 26 02:25:54 2017]
ElapsedTime:8 sec
RETURN rc = 0
OUTPUT:
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
5 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
tz003: Error: No archictecture defined in nodetype table for the node.
tz004: Error: serialport defined, but no serialspeed for this node in nodehm table
Error: Failed to generate xnba configurations for some node(s) on c910f03c01p09. Check xCAT log file for more details.
tz001: shell
tz002: shell
tz005: shell
tz003: Error: No archictecture defined in nodetype table for the node.
tz004: Error: serialport defined, but no serialspeed for this node in nodehm table
tz001: offline
tz002: offline
tz003: offline
tz004: offline
tz005: offline
5 object definitions have been removed.
Error: Could not find an object named 'rhels7.99-ppc64le-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels7.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels6.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
1 object definitions have been removed.
CHECK:rc == 0	[Pass]

------END::nodeset_shell_xnba::Passed::Time:Thu Oct 26 02:26:02 2017 ::Duration::8 sec------
------START::nodeset_osimage_grub2::Time:Thu Oct 26 02:26:02 2017------

RUN:/opt/xcat/share/xcat/tools/autotest/testcase/nodeset/nodeset_osimage_grub2 [Thu Oct 26 02:26:02 2017]
ElapsedTime:9 sec
RETURN rc = 0
OUTPUT:
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
tz002: Error: No MAC address available for this node
Error: Failed to generate grub2 configurations for some node(s) on c910f03c01p09. Check xCAT log file for more details.
tz001: install rhels7.99-ppc64-compute
tz003: install rhels7.99-ppc64-compute
tz004: install rhels7.99-ppc64-compute
tz005: install rhels7.99-ppc64-compute
tz002: Error: No MAC address available for this node
tz001: offline
tz002: offline
tz003: offline
tz004: offline
tz005: offline
5 object definitions have been removed.
Error: Could not find an object named 'rhels7.99-ppc64le-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
1 object definitions have been removed.
Error: Could not find an object named 'rhels6.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels6.99-x86_64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
CHECK:rc == 0	[Pass]

------END::nodeset_osimage_grub2::Passed::Time:Thu Oct 26 02:26:11 2017 ::Duration::9 sec------
------START::nodeset_osimage_petitboot::Time:Thu Oct 26 02:26:11 2017------

RUN:/opt/xcat/share/xcat/tools/autotest/testcase/nodeset/nodeset_osimage_petitboot [Thu Oct 26 02:26:11 2017]
ElapsedTime:9 sec
RETURN rc = 0
OUTPUT:
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
5 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
tz002: Error: No MAC address available for this node
tz004: Error: serialport defined, but no serialspeed for this node in nodehm table
Error: Failed to generate petitboot configurations for some node(s) on c910f03c01p09. Check xCAT log file for more details.
tz001: install rhels7.99-ppc64le-compute
tz003: install rhels7.99-ppc64le-compute
tz005: install rhels7.99-ppc64le-compute
tz002: Error: No MAC address available for this node
tz004: Error: serialport defined, but no serialspeed for this node in nodehm table
tz001: offline
tz002: offline
tz003: offline
tz004: offline
tz005: offline
5 object definitions have been removed.
1 object definitions have been removed.
Error: Could not find an object named 'rhels7.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels6.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels6.99-x86_64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
CHECK:rc == 0	[Pass]

------END::nodeset_osimage_petitboot::Passed::Time:Thu Oct 26 02:26:20 2017 ::Duration::9 sec------
------START::nodeset_osimage_xnba::Time:Thu Oct 26 02:26:20 2017------

RUN:/opt/xcat/share/xcat/tools/autotest/testcase/nodeset/nodeset_osimage_xnba [Thu Oct 26 02:26:20 2017]
ElapsedTime:9 sec
RETURN rc = 0
OUTPUT:
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
5 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
tz002: Error: No MAC address available for this node
tz004: Error: serialport defined, but no serialspeed for this node in nodehm table
Error: Failed to generate xnba configurations for some node(s) on c910f03c01p09. Check xCAT log file for more details.
tz001: install rhels6.99-x86_64-compute
tz003: install rhels6.99-x86_64-compute
tz005: install rhels6.99-x86_64-compute
tz002: Error: No MAC address available for this node
tz004: Error: serialport defined, but no serialspeed for this node in nodehm table
tz001: offline
tz002: offline
tz003: offline
tz004: offline
tz005: offline
5 object definitions have been removed.
Error: Could not find an object named 'rhels7.99-ppc64le-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels7.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels6.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
1 object definitions have been removed.
CHECK:rc == 0	[Pass]

------END::nodeset_osimage_xnba::Passed::Time:Thu Oct 26 02:26:29 2017 ::Duration::9 sec------
------START::nodeset_nonexistent_osimage_grub2::Time:Thu Oct 26 02:26:29 2017------

RUN:/opt/xcat/share/xcat/tools/autotest/testcase/nodeset/nodeset_nonexistent_osimage_grub2 [Thu Oct 26 02:26:29 2017]
ElapsedTime:10 sec
RETURN rc = 0
OUTPUT:
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
tz001: install rhels7.99-ppc64-compute
tz002: install rhels7.99-ppc64-compute
tz003: install rhels7.99-ppc64-compute
tz004: install rhels7.99-ppc64-compute
tz005: install rhels7.99-ppc64-compute
Error: Cannot find the OS image nonexistent-osimage on the osimage table.
Error: Cannot find the OS image nonexistent-osimage on the osimage table.
tz001: offline
tz002: offline
tz003: offline
tz004: offline
tz005: offline
1 object definitions have been created or modified.
tz003: Error: Cannot find the OS image nonexistent-osimage on the osimage table
Error: Failed to generate grub2 configurations for some node(s) on c910f03c01p09. Check xCAT log file for more details.
tz001: install rhels7.99-ppc64-compute
tz002: install rhels7.99-ppc64-compute
tz004: install rhels7.99-ppc64-compute
tz005: install rhels7.99-ppc64-compute
tz003: Error: Cannot find the OS image nonexistent-osimage on the osimage table
tz001: offline
tz002: offline
tz003: offline
tz004: offline
tz005: offline
5 object definitions have been removed.
Error: Could not find an object named 'rhels7.99-ppc64le-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
1 object definitions have been removed.
Error: Could not find an object named 'rhels6.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels6.99-x86_64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
CHECK:rc == 0	[Pass]

------END::nodeset_nonexistent_osimage_grub2::Passed::Time:Thu Oct 26 02:26:39 2017 ::Duration::10 sec------
------START::nodeset_nonexistent_osimage_petitboot::Time:Thu Oct 26 02:26:39 2017------

RUN:/opt/xcat/share/xcat/tools/autotest/testcase/nodeset/nodeset_nonexistent_osimage_petitboot [Thu Oct 26 02:26:39 2017]
ElapsedTime:13 sec
RETURN rc = 0
OUTPUT:
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
tz001: install rhels7.99-ppc64le-compute
tz002: install rhels7.99-ppc64le-compute
tz003: install rhels7.99-ppc64le-compute
tz004: install rhels7.99-ppc64le-compute
tz005: install rhels7.99-ppc64le-compute
Error: Cannot find the OS image nonexistent-osimage on the osimage table.
Error: Cannot find the OS image nonexistent-osimage on the osimage table.
tz001: offline
tz002: offline
tz003: offline
tz004: offline
tz005: offline
1 object definitions have been created or modified.
tz003: Error: Cannot find the OS image nonexistent-osimage on the osimage table
Error: Failed to generate petitboot configurations for some node(s) on c910f03c01p09. Check xCAT log file for more details.
tz001: install rhels7.99-ppc64le-compute
tz002: install rhels7.99-ppc64le-compute
tz004: install rhels7.99-ppc64le-compute
tz005: install rhels7.99-ppc64le-compute
tz003: Error: Cannot find the OS image nonexistent-osimage on the osimage table
tz001: offline
tz002: offline
tz003: offline
tz004: offline
tz005: offline
1 object definitions have been created or modified.
tz003: Error: Cannot find the OS image nonexistent-osimage on the osimage table
Error: Failed to generate petitboot configurations for some node(s) on c910f03c01p09. Check xCAT log file for more details.
tz001: install rhels7.99-ppc64le-compute
tz002: install rhels7.99-ppc64le-compute
tz004: install rhels7.99-ppc64le-compute
tz005: install rhels7.99-ppc64le-compute
tz001: offline
tz002: offline
tz003: offline
tz004: offline
tz005: offline
5 object definitions have been removed.
1 object definitions have been removed.
Error: Could not find an object named 'rhels7.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels6.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels6.99-x86_64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
CHECK:rc == 0	[Pass]

------END::nodeset_nonexistent_osimage_petitboot::Passed::Time:Thu Oct 26 02:26:52 2017 ::Duration::13 sec------
------START::nodeset_nonexistent_osimage_xnba::Time:Thu Oct 26 02:26:52 2017------

RUN:/opt/xcat/share/xcat/tools/autotest/testcase/nodeset/nodeset_nonexistent_osimage_xnba [Thu Oct 26 02:26:52 2017]
ElapsedTime:10 sec
RETURN rc = 0
OUTPUT:
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
1 object definitions have been created or modified.
tz001: install rhels6.99-x86_64-compute
tz002: install rhels6.99-x86_64-compute
tz003: install rhels6.99-x86_64-compute
tz004: install rhels6.99-x86_64-compute
tz005: install rhels6.99-x86_64-compute
Error: Cannot find the OS image nonexistent-osimage on the osimage table.
Error: Cannot find the OS image nonexistent-osimage on the osimage table.
tz001: offline
tz002: offline
tz003: offline
tz004: offline
tz005: offline
1 object definitions have been created or modified.
tz003: Error: Cannot find the OS image nonexistent-osimage on the osimage table
Error: Failed to generate xnba configurations for some node(s) on c910f03c01p09. Check xCAT log file for more details.
tz001: install rhels6.99-x86_64-compute
tz002: install rhels6.99-x86_64-compute
tz004: install rhels6.99-x86_64-compute
tz005: install rhels6.99-x86_64-compute
tz003: Error: Cannot find the OS image nonexistent-osimage on the osimage table
tz001: offline
tz002: offline
tz003: offline
tz004: offline
tz005: offline
5 object definitions have been removed.
Error: Could not find an object named 'rhels7.99-ppc64le-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels7.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
Error: Could not find an object named 'rhels6.99-ppc64-install-compute' of type 'osimage'.
No objects have been removed from the xCAT database.
1 object definitions have been removed.
CHECK:rc == 0	[Pass]

------END::nodeset_nonexistent_osimage_xnba::Passed::Time:Thu Oct 26 02:27:02 2017 ::Duration::10 sec------
------Total: 9 , Failed: 0------

xCAT automated test finished atThu Oct 26 02:27:02 2017
Please check results in the /opt/xcat/bin/../share/xcat/tools/autotest/result/,
and see /opt/xcat/bin/../share/xcat/tools/autotest/result//failedcases.20171026022534 file for failed cases.
see /opt/xcat/bin/../share/xcat/tools/autotest/result//performance.report.20171026022534 file for time consumption
```

**Pull request merged**